exchanges/stream/websocket.go

```diff
@@ -845,23 +849,23 @@ func (w *Websocket) GetName() string {
 // GetChannelDifference finds the difference between the subscribed channels
 // and the new subscription list when pairs are disabled or enabled.
 func (w *Websocket) GetChannelDifference(genSubs []ChannelSubscription) (sub, unsub []ChannelSubscription) {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
+	w.subscriptionMutex.RLock()
+	defer w.subscriptionMutex.RUnlock()
 
 oldsubs:
-	for x := range w.subscriptions {
+	for _, x := range w.subscriptions {
 		for y := range genSubs {
-			if w.subscriptions[x].Equal(&genSubs[y]) {
+			if x.Equal(&genSubs[y]) {
 				continue oldsubs
 			}
 		}
-		unsub = append(unsub, w.subscriptions[x])
+		unsub = append(unsub, x)
 	}
 
 newsubs:
 	for x := range genSubs {
-		for y := range w.subscriptions {
-			if genSubs[x].Equal(&w.subscriptions[y]) {
+		for _, y := range w.subscriptions {
+			if genSubs[x].Equal(&y) { //nolint:gosec // for alias var is not closured or stored
```

Comment:
Yeah. I really don't know. I even looked at the source for gosec to see if there were mitigations available to us.

Reply:
Building upon the earlier theme, I'll be more intrusive on this and everything will be better.
I felt a psychological reluctance because I haven't tangled with this func much and I saw that if I touched it I'd have to rewrite it 🤦 

Pretty happy the simplicity of the rewrite though; Hope you like it!
Added slightly better test coverage to be sure.

Fixed with b23b68fe

Reply:
@gloriousCode Note: My solution is non-deterministic for unsubs, revealed by transient failures of my new test.
I think that's okay though, so I'll fix my test.